intern/cycles/kernel/kernel_color.h

```diff
@@ -33,6 +33,673 @@ ccl_device float linear_rgb_to_gray(KernelGlobals *kg, float3 c)
+        make_float4(797.0, 0.000012654360f, 0.000004569720f, 0.000000000000f),
+        make_float4(798.0, 0.000011797230f, 0.000004260194f, 0.000000000000f),
+        make_float4(799.0, 0.000010998440f, 0.000003971739f, 0.000000000000f),
+        make_float4(800.0, 0.000010253980f, 0.000003702900f, 0.000000000000f),
+        make_float4(801.0, 0.000009559646f, 0.000003452163f, 0.000000000000f),
+        make_float4(802.0, 0.000008912044f, 0.000003218302f, 0.000000000000f),
+        make_float4(803.0, 0.000008308358f, 0.000003000300f, 0.000000000000f),
+        make_float4(804.0, 0.000007745769f, 0.000002797139f, 0.000000000000f),
+        make_float4(805.0, 0.000007221456f, 0.000002607800f, 0.000000000000f),
+        make_float4(806.0, 0.000006732475f, 0.000002431220f, 0.000000000000f),
+        make_float4(807.0, 0.000006276423f, 0.000002266531f, 0.000000000000f),
+        make_float4(808.0, 0.000005851304f, 0.000002113013f, 0.000000000000f),
+        make_float4(809.0, 0.000005455118f, 0.000001969943f, 0.000000000000f),
+        make_float4(810.0, 0.000005085868f, 0.000001836600f, 0.000000000000f),
+        make_float4(811.0, 0.000004741466f, 0.000001712230f, 0.000000000000f),
+        make_float4(812.0, 0.000004420236f, 0.000001596228f, 0.000000000000f),
+        make_float4(813.0, 0.000004120783f, 0.000001488090f, 0.000000000000f),
+        make_float4(814.0, 0.000003841716f, 0.000001387314f, 0.000000000000f),
+        make_float4(815.0, 0.000003581652f, 0.000001293400f, 0.000000000000f),
+        make_float4(816.0, 0.000003339127f, 0.000001205820f, 0.000000000000f),
+        make_float4(817.0, 0.000003112949f, 0.000001124143f, 0.000000000000f),
+        make_float4(818.0, 0.000002902121f, 0.000001048009f, 0.000000000000f),
+        make_float4(819.0, 0.000002705645f, 0.000000977058f, 0.000000000000f),
+        make_float4(820.0, 0.000002522525f, 0.000000910930f, 0.000000000000f),
+        make_float4(821.0, 0.000002351726f, 0.000000849251f, 0.000000000000f),
+        make_float4(822.0, 0.000002192415f, 0.000000791721f, 0.000000000000f),
+        make_float4(823.0, 0.000002043902f, 0.000000738090f, 0.000000000000f),
+        make_float4(824.0, 0.000001905497f, 0.000000688110f, 0.000000000000f),
+        make_float4(825.0, 0.000001776509f, 0.000000641530f, 0.000000000000f),
+        make_float4(826.0, 0.000001656215f, 0.000000598090f, 0.000000000000f),
+        make_float4(827.0, 0.000001544022f, 0.000000557575f, 0.000000000000f),
+        make_float4(828.0, 0.000001439440f, 0.000000519808f, 0.000000000000f),
+        make_float4(829.0, 0.000001341977f, 0.000000484612f, 0.000000000000f),
+        make_float4(830.0, 0.000001251141f, 0.000000451810f, 0.000000000000f)
+    };
+    return find_position_in_lookup_unit_step(wavelength_xyz_lookup, wavelength, 360, 830);
+}
+
+ccl_device float3 wavelength_intensities_to_linear(KernelGlobals *kg, float3 intensities, float3 wavelengths)
```

Comment:
This is the function that is used to actually convert the 3 wavelengths for a ray into XYZ data before it is added to the film. Since light is additive (when working with a linear representation of intensity), we can just sum the XYZ coordinates to get the result for the three wavelengths added together.